Implement new concept with exercise: randomness (#2139)
exercism/go: concepts/randomness/about.md

> # Randomness Introduction

Reviewer: ```suggestion
# About
```

> Package [`math\rand`][mathrand] provides support for generating pseudo-random numbers.

Reviewer: ```suggestion
Package [math\rand][mathrand] provides support for generating pseudo-random numbers.
```

Although the intention is good, my experience is that the block style might not make it obvious that the text is a link, so it's better to remove it.

Can you fix this in the other places too please?

Author: Thanks - done.

Here is how to generate a random integer between `0` and `99`:

```go
n := rand.Intn(100) // n is a random int, 0 <= n < 100
```

Function `rand.Float64` returns a random floating point number between `0.0` and `1.0`:

```go
f := rand.Float64() // f is a random float64, 0.0 <= f < 1.0
```

There is also support for shuffling a slice (or other data structures):

```go
x := []string{"a", "b", "c", "d", "e"}
// shuffling the slice put its elements into a random order
rand.Shuffle(len(x), func(i, j int) {
	x[i], x[j] = x[j], x[i]
})
```

## Seeds

The number sequences generated by package `math\rand` are not truly random.  
They are completely determined by an initial `seed` value.
By default, the package uses the number `1` as a seed.

In order to get a different sequence of random numbers, we need to set a different seed.
A simple approach is to seed with the current computer time.

```go
rand.Seed(time.Now().UnixNano())
```

> Seeding must be done **before** any of the functions that generate random numbers.

Reviewer: ```suggestion
Setting the seed must be done **before** calling any of the functions that generate random numbers. We usually only need to set the seed once in our program. From one seed we can generate many random numbers.
```

Additional clarification. Also apply this change in the other places where this is present.

Author: Done :)

### Warning

If you don't call `rand.Seed`, then the package will produce the same sequence of random numbers each time your program runs!

As an example, consider the following program:

```go
package main

import (
	"fmt"
	"math/rand"
)

func main() {
    // rand.Seed(time.Now().UnixNano())
	a := rand.Intn(1000000)
	b := rand.Intn(1000000)
	c := rand.Intn(1000000)
	fmt.Println(a, b, c)
}
```

Every time this program runs, it will produce the same output:

```text
498081 727887 131847
```

Once we uncomment the first line in the `main()` function, the output varies between different runs:

```text
817603 745216 422387

265737 21561 640434

708506 627509 363987
```

> ### Caution
> 
> The output produced by package `math/rand` might be easily predictable regardless of how it is seeded.
> For random numbers suitable for security-sensitive work, you should use the [`crypto/rand`][cryptorand] package.

Reviewer: ```suggestion
~~~~exercism/caution

The output produced by package `math/rand` might be easily predictable regardless of how it is seeded.
For random numbers suitable for security-sensitive work, you should use the [crypto/rand][cryptorand] package.
~~~~
```

We can take advantage of this special syntax exercism allows to highlight this block. To read more about this check [markdown](https://exercism.org/docs/building/markdown/markdown) in the docs.

Apply this change in the other places where this is present.

Author: This caution is currently only in file `about.md`. Should it maybe also go in the concept introduction in the randomness folder?

Reviewer: You are right, thought I've seen this somewhere else, disregard the advice then.

### Advanced

Instead of seeding with computer time, it is better to use `crypt/rand` functions to generate a seed for `math/rand`.

> For more information, see this [stackoverflow article][stackoverflow].

Reviewer: I think there should be a brief explanation on why `crypt/rand` is "better". In what sense? Following the stack overflow link is also not clear for me what you mean. Maybe there's a better resource for it?

Author: In essence, using `time` as a seed means a lot of the bits in the seed will end up the same, especially when the program is run several times in short succession and computer time lacks precision. Using a `crypt/rand` generated seed gives more entropy, hence less predictable (and potentially less brute-forceable) random-numbers sequences generated by `math/rand`. So in principle, this looks like sound advice. Unfortunately, I have not been able to find a better resource to link to. I suggest we remove this section for now and keep it on a "under consideration" list for the future.

Reviewer: I think we can remove it too. We can just mention that crypt/rand can be used for randomly secure cryptography applications and leave it at that.

Maybe explaining the difference between the packages would be more suitable in a future exercise where we talk about creating generators from different sources.
